### PR TITLE
feat(DefaultItinerary): show CO2 for each trip option

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
TODO: Update core-utils version from https://github.com/opentripplanner/otp-ui/pull/346

This update adds CO2 emissions to the default itinerary view, and some new configuration options associated with it. Ensure you have https://github.com/opentripplanner/otp-ui/pull/346 in your modules folder!

### CODE REVIEW

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:340]: Seeing NaN for bicycle_rent -- we probably should have a better null check and also include a value for bike_rent 

Screen Shot 2022-02-09 at 7 48 55 AM

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:146]: Why do we need the noSpace? The icon looks the same as the others

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:157]: Would it be possible to add the feature we were talking about where the value is only shown if there is a significant difference (and then making the results really green and really red perhaps based on how big of a difference there is)? These rows are getting really tall... and 15oz vs 14oz is not always the most helpful info...

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:163]: Should we maybe consider adding some rounding? The nearest 15 or 30 or even 50 might make the values a bit more glancable

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:336]: I would move all the carbonIntensity values to core-utils and then make massUnit the only valid setting and then we can use typescript to enforce these options

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
A good start! I love the leaf. Just a few small implementation questions. Also, I'm not seeing the values in tabbed mode -- am I missing something?

File: example-config.yml
Comment (daniel-heppner-ibigroup) [reply example-config.yml:336]: You don't think the agencies might want to customize the carbon intensity of each mode?

File: example-config.yml
Comment (daniel-heppner-ibigroup) [reply example-config.yml:340]: I thought I fixed that!

File: lib/components/narrative/default/default-itinerary.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js:146]: There was a margin around LegIconWrapper and I wasn't sure why, so instead of removing it I just added this prop.

File: lib/components/narrative/default/default-itinerary.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js]: We have the percentages/relative values, as well as some rounding and hiding values below 10%. Does this feel like enough? The design is not great, but it really just needs a complete redesign. I think adding bright colors would be a bit too much.

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:337]: we should add `enabled: true` here!

File: i18n/en-US.yml
Comment (miles-grant-ibigroup) [inline i18n/en-US.yml:775]: I would shorten this string so that the time isn't pushed awkwardly to the middle. Maybe `XX% of car emissions`? Either that or we should figure out a way to force a line break

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:165]: Is this auto formatted?

File: lib/components/narrative/narrative-itineraries.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/narrative-itineraries.js:191]: oh no looks like this snuck in!

File: i18n/en-US.yml
Comment (miles-grant-ibigroup) [reply i18n/en-US.yml:775]: We should also be careful with language. if a car emits 100 grams, and we say 85% less co2 then isn't that 15 grams? is that what we mean or do we mean 85 grams

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Looking really good well done! As much as I think we should re-work the batch itinerary list I think having the time pushed to the middle is problematic which is the main thing.

Also, I'm not seeing the co2 results in trip-details even though I've updated trip-details to your version. Not sure what's going on there!

File: lib/components/narrative/default/default-itinerary.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js:165]: supposedly it was... I added some spaces

File: i18n/en-US.yml
Comment (daniel-heppner-ibigroup) [reply i18n/en-US.yml:775]: I think having the word "less" and "more" is pretty important to make it clear to understand at a glance.

Comment (daniel-heppner-ibigroup) [reply i18n/en-US.yml:775]: OK I aligned all the attributes to the right. Pretty big change, but I think I do like it more. Let me know what you think.

Comment (miles-grant-ibigroup) [reply i18n/en-US.yml:775]: This definitely looks better thanks! Feels a bit busy but we can probably leave that for the redesign

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:350]: Optional chaining is probably needed here!

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Looking really good, although I'm going go request changes since I'm seeing some white screen of death errors with the TripDetails. I guess we have to leave this blocked for now

File: lib/components/narrative/default/default-itinerary.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js:350]: Hmmm... I don't think so. If CO2 was null then it would be crashing on the previous line, right?

Comment (miles-grant-ibigroup) [reply lib/components/narrative/default/default-itinerary.js:350]: Are we ever checking for the `co2` field of `config`?

Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js:350]: It should be checked everywhere that it's used. You did catch a place in the other PR that it wasn't being checked, though, so maybe that was causing the error? That's fixed now.

**Comment by miles-grant-ibigroup** [comment]
I think I found the reason the Percy tests are failing: Screen Shot 2022-08-15 at 7 18 33 AM

As soon as I plan a route the app crashes. My guess is it has something to do with mismatched otp-ui versions? I tried upgrading a few to their latest but that only changed the message I got. I think @daniel-heppner-ibigroup you know better than me which package versions are needed.

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:379]: Are these still accurate? I'm not seeing them have an effect on the results. Additionally, the defaults seem a bit strange: for a certain itinerary I'm getting that walking is 85% of driving alone, but scooter is 100% less? That doesn't make any sense

File: lib/components/narrative/default/default-itinerary.js
Comment (miles-grant-ibigroup) [inline lib/components/narrative/default/default-itinerary.js:146]: Should this maybe be configurable? I would also set this to something like 33% or something so it doesn't show up as often

File: lib/components/narrative/metro/metro-itinerary.tsx
Comment (miles-grant-ibigroup) [inline lib/components/narrative/metro/metro-itinerary.tsx:271]: I don't like the `{' '}`. Would it be possible to insert the number as a value inside the message and then pass this entire component in as the value?

File: lib/components/narrative/metro/metro-itinerary.tsx
Comment (miles-grant-ibigroup) [inline lib/components/narrative/metro/metro-itinerary.tsx:263]: Is this 20 related to the threshold we saw previously?

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Looks really slick! A few small questions but I think this is almost good to go!

Also, this seems to not work with the "old" batch ui or the tabbed UI. Is this intentional?

File: example-config.yml
Comment (daniel-heppner-ibigroup) [reply example-config.yml:379]: The mass units should show up in the trip details, where the actual mass of the CO2 is shown. It won't have any effect on the percentages, of course.

File: lib/components/narrative/default/default-itinerary.js
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/default/default-itinerary.js]: Done!

File: lib/components/narrative/metro/metro-itinerary.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/metro/metro-itinerary.tsx]: uhhh maybe. it's gone now, uses the config!

**Comment by daniel-heppner-ibigroup** [comment]
You should be able to see this on the old batch UI! 
image

File: example-config.yml
Comment (daniel-heppner-ibigroup) [reply example-config.yml:379]: Hm there does appear to be an issue where the units are not changing. I think this is a problem on the OTP-UI side, since the entire CO2 config just gets passed into TripDetails.

Comment (daniel-heppner-ibigroup) [reply example-config.yml:379]: And finally, can you give me an example itinerary where you're seeing the wacky numbers? That sounds weird.

Comment (miles-grant-ibigroup) [reply example-config.yml:379]: I'm seeing it on all e-scooter share itineraries. My guess is it may be related to the name changing between `SCOOTER` to `MICROMOBILITY_RENT`? In that case I think our fallback should be infinite carbon rather than 0 carbon

Comment (miles-grant-ibigroup) [reply example-config.yml:379]: I've now discovered this was a configuration issue. Thanks for the new documentation!

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:398]: Is it possible to rename this to `showIfHigher`? Since it's already a child of `co2`

File: example-config.yml
Comment (miles-grant-ibigroup) [inline example-config.yml:399]: Is it possible to rename this to `cutoff`? Since it's already a child of `co2`

File: lib/components/narrative/metro/metro-itinerary.tsx
Comment (miles-grant-ibigroup) [inline lib/components/narrative/metro/metro-itinerary.tsx:266]: Bit of a nitpick, but the `enabled` check can probably come first

Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/metro-itinerary.tsx:266]: Additionally, do we need an explicit `enabled` flag? Can't we just use the presence of one or more `carbonIntensity` keys?

**Review by miles-grant-ibigroup — APPROVED** [review]
A few small nitpicks, but overall I think this looks great! Thanks so much for all your work on this I know it's been a long process

File: lib/components/narrative/metro/metro-itinerary.tsx
Comment (daniel-heppner-ibigroup) [reply lib/components/narrative/metro/metro-itinerary.tsx:266]: Not really, because someone (probably most people) will just want to use the defaults. Technically we could check if the `co2` property exists in the config but I think that would look a little weird. You'd have the property like `co2:` in the yaml all on its own with nothing under it.

Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/metro-itinerary.tsx]: That's a good point. I'll leave it up to the second reviewer to decide on this!

**Comment by demory** [comment]
Quick question: am I supposed to see CO2 numbers in the trip summary for walk/bike alone trips? I see the "X% lower than driving" bar at the top of these itineraries but nothing in the summary.

**Comment by daniel-heppner-ibigroup** [comment]
@demory The problem appeared to be the outdated `trip-details` package. Not sure why that never got updated, but I've changed it to the latest version which seems to have fixed the issue!

**Comment by daniel-heppner-ibigroup** [comment]
@miles-grant-ibigroup Do you know why all the times switched from AM to PM in the percy snapshots?

**Comment by miles-grant-ibigroup** [comment]
@daniel-heppner-ibigroup The am/pm thing is a really old bug from when we were adjusting date formatting! If you merge latest dev into this branch it should go away.

**Comment by demory** [comment]
@daniel-heppner-ibigroup to recap what we discussed on our last call:
- I am still seeing the issue with no CO2 data in the trip details box for walk/bike only trips. The bug you addressed above appears to be different
- We talked about not showing CO2 at all for those trips (both the header bar and the details box entry), or having that be configurable

**Comment by daniel-heppner-ibigroup** [comment]
I've made some changes to iron out a bunch of bugs and also improve the flow of data. This requires a change to OTP-UI as well, which I've done in a separate PR here. https://github.com/opentripplanner/otp-ui/pull/461
We will need to update the version numbers in this PR once that one is merged in.

**Comment by daniel-heppner-ibigroup** [comment]
@demory This is ready for a review by the way!

**Comment by demory** [comment]
Just need to implement `showIfHigher` in the metro UI and this should be good to go

**Comment by daniel-heppner-ibigroup** [comment]
Good to go, @demory. I fixed up the issues with `showIfHigher`

**Review by demory — APPROVED** [review]

**Pull request merged**
